README.md
# MyTasks

[![PSGallery Version](https://img.shields.io/powershellgallery/v/myTasks.png?style=for-the-badge&logo=powershell&label=PowerShell%20Gallery)](https://www.powershellgallery.com/packages/myTasks/) [![PSGallery Downloads](https://img.shields.io/powershellgallery/dt/MyTasks.png?style=for-the-badge&label=Downloads)](https://www.powershellgallery.com/packages/MyTasks/)

This PowerShell module is designed as a task or a simple To-Do manager. The module contains several commands for working with tasks. It should work with both Windows PowerShell and PowerShell 7.x  with a few limitations. You can install the latest version from the PowerShell Gallery. You might want to need the `-Scope`parameter for PowerShell 7.x.

```powershell
Install-Module MyTasks [-scope currentuser]
```

Task data is stored in an XML file. Other configuration information is stored in simple text files. Here are a few highlights.

## A Class-Based Module

This module uses a class definition for the task object and is designed to work on both Windows PowerShell and PowerShell Core.

```powershell
Class MyTask {
    [int]$ID
    [string]$Name
    [string]$Description
    [datetime]$DueDate
    [bool]$Overdue
    [String]$Category
    [ValidateRange(0, 100)][int]$Progress
    hidden[bool]$Completed
    hidden[datetime]$TaskCreated = (Get-Date)
    hidden[datetime]$TaskModified
    hidden[guid]$TaskID = (New-Guid)
    # ID and OverDue values are calculated at run time.

    # Methods

    #set task as completed
    [void]CompleteTask([datetime]$CompletedDate) {
        Write-Verbose "[CLASS  ] Completing task: $($this.name)"
        $this.Completed = $True
        $this.Progress = 100
        $this.Overdue = $False
        $this.TaskModified = $CompletedDate
    }

    #check if task is overdue and update
    hidden [void]Refresh() {
        Write-Verbose "[CLASS  ] Refreshing task $($this.name)"
        #only mark as overdue if not completed and today is greater than the due date
        Write-Verbose "[CLASS  ] Comparing $($this.DueDate) due date to $(Get-Date)"

        if ($This.completed) {
            $this.Overdue = $False
        }
        elseif ((Get-Date) -gt $this.DueDate) {
            $this.Overdue = $True
        }
        else {
            $this.Overdue = $False
        }

    } #refresh

    #Constructors
    MyTask([string]$Name) {
        Write-Verbose "[CLASS  ] Constructing with name: $name"
        $this.Name = $Name
        $this.DueDate = (Get-Date).AddDays(7)
        $this.TaskModified = (Get-Date)
        $this.Refresh()
    }
    #used for importing from XML
    MyTask([string]$Name, [datetime]$DueDate, [string]$Description, [string]$Category, [boolean]$Completed) {
        Write-Verbose "[CLASS  ] Constructing with due date, description and category"
        $this.Name = $Name
        $this.DueDate = $DueDate
        $this.Description = $Description
        $this.Category = $Category
        $this.TaskModified = $this.TaskCreated
        $this.Completed = $completed
        $this.Refresh()
    }

} #end class definition
```

While you could use the object's properties and methods directly, you should use the appropriate module command.

## XML Data

All of the task information is stored in an XML file. The commands in this module will read in, update, and remove items as needed using PowerShell commands such as `Select-XML`. By default, these files are stored in your Documents folder (on Windows systems) or in the Home folder (on Linux). You can change the default location by using the [Set-myTaskHome](docs/Set-MyTaskHome.md) command. This is helpful if you are sharing task information between computers via a service like Dropbox or OneDrive.

```powershell
Set-MyTaskHome -path C:\Users\Jeff\dropbox\mytasks\
```

If you use this feature, you'll need to make sure you run this command before doing anything. It is recommended to put this command in a PowerShell profile script.

You shouldn't have to manage the module related variables directly. Use `Get-MyTaskHome` to view your current settings.

## Categories

The Task object includes a Category property. The module will define a default set of categories ("Work","Personal","Customer","Other"), but users can create their own by using the MyTaskCategory commands:

+ [Add-MyTaskCategory](docs/Add-MyTaskCategory.md)
+ [Get-MyTaskCategory](docs/Get-MyTaskCategory.md)
+ [Remove-MyTaskCategory](docs/Remove-MyTaskCategory.md)

Task information is stored in a text file in the `$myTaskHome` location.

## Basic Usage

You create a task with at least a name and category. The default due date will be 7 days from the current date and time.

```powershell
New-MyTask "return library books" -Category personal
```

You can also specify a due date.

```powershell
New-MyTask "Pluralsight" -duedate "2/1/2020" -description "renew subscription" -category other
```

You can use `Set-MyTask` to modify a task.

```powershell
Get-MyTask Pluralsight | Set-Mytask -DueDate 3/1/2020
```

Because the task has a Progress property, you can use [Set-MyTask](docs/Set-MyTask.md) to update that as well. This is intended to be used as a percentage complete.

```powershell
Set-Mytask "book review" -Progress 60
```

Use `Get-MyTask` to view tasks. Normally, you will use [Get-MyTask](docs/Get-MyTask.md) to display all tasks, some tasks or a single item:

```powershell
PS S:\> Get-MyTask -name MemoryTools

ID  Name         Description                DueDate OverDue Category  Progress
--  ----         -----------                ------- ------- --------  --------
8   MemoryTools  update module            7/22/2020 False   Projects        10
```

The default behavior is to display incomplete tasks due in the next 30 days. Look at the help for `Get-MyTask` for more information.

This module has a custom format file that uses ANSI escape sequences to color-code tasks. Overdue tasks will be shown in red. Tasks due in the next 48 hours will be shown in orange.

![Get-MyTask](images/get-mytask.png)

This command may not work as expected in the PowerShell ISE or the VS Code PowerShell Integrated shell.

When a task is finished, you can mark it as complete.

```powershell
Complete-MyTask -name "order coffee"
```

The task will remain but be marked as 100% complete. You can still see the task when using the -All parameter with `Get-MyTask` or `Show-MyTask`. At some point you might want to remove completed tasks from the master XML file. You can use [Remove-MyTask](docs/Remove-MyTask.md) to permanently delete them. Or use the `Archive-MyTask` command to move them to an archive XML file.

## Format Views

The module includes a `format.ps1xml` file that defines a default display when you run `Get-MyTask`. You will get a slightly different set of properties when you run `Get-MyTask | Format-List`. There is also a custom table view called `Category` which will create a table grouped by the Category property. You should sort the tasks first:

```powershell
Get-MyTask | Sort-Object Category | Format-Table -view category
```

Or you can use the `DueDate` table view - sort on DueDate

```powershell
Get-MyTask -days 180 | Sort-Object duedate | Format-Table -view duedate
```

![formatted views](images/show-mytask-2.png)

## Archiving and Removing

Over time your task file might get quite large. Even though the default behavior is to ignore completed tasks, you have an option to archive them to a separate XML file using `Save-MyTask` which has an alias of `Archive-MyTask`:

```powershell
Get-MyTask -Completed | Archive-MyTask
```

There is an option to archive tasks when you run [Complete-MyTask](docs/Complete-MyTask.md). Or you can completely delete a task with `Remove-MyTask`.

Use the `Get-MyTaskArchive` to view archived tasks.

## Email Reminders

If you are running this module on Windows PowerShell that includes the PSScheduledJob module, you can create a scheduled PowerShell job that will send you a daily email with tasks that are due in 3 days or less. The default is a plain text message but you can also send it as HTML. Use the [Enable-EmailReminder](docs/Enable-EmailReminder.md) command to set up the job.

You should read full help and examples for all commands as well as the [about_MyTasks](docs/about_MyTasks.md) help file.

+ [Add-MyTaskCategory](docs/Add-MyTaskCategory.md)
+ [Backup-MyTaskFile](docs/Backup-MyTaskFile.md)
+ [Complete-MyTask](docs/Complete-MyTask.md)
+ [Get-MyTask](docs/Get-MyTask.md)
+ [Get-MyTaskCategory](docs/Get-MyTaskCategory.md)
+ [New-MyTask](docs/New-MyTask.md)
+ [Remove-MyTask](docs/Remove-MyTask.md)
+ [Remove-MyTaskCategory](docs/Remove-MyTaskCategory.md)
+ [Save-MyTask](docs/Save-MyTask.md)
+ [Set-MyTask](docs/Set-MyTask.md)
+ [Enable-EmailReminder](docs/Enable-EmailReminder.md)
+ [Disable-EmailReminder](docs/Disable-EmailReminder.md)
+ [Get-EmailReminder](docs/Get-EmailReminder.md)
+ [Set-MyTaskHome](docs/Set-MyTaskHome.md)
+ [Get-MyTaskHome](docs/Get-MyTaskHome.md)
+ [Get-MyTaskArchive](docs/Get-MyTaskArchive.md)

## Limitations

Please post any issues, questions, or feature requests in the [Issues](https://github.com/jdhitsolutions/MyTasks/issues) section of this module's GitHub repository.

Last Updated 2020-10-14 16:00:30Z


changelog.md
# Change Log for MyTasks

## v2.4.0

+ Restructured module layout
+ Removed `Show-MyTask` and modified default format view for `Get-MyTask` to use ANSI escape sequences instead of `Write-Host`. (Issue #44) *Breaking Change*
+ Added the aliases `Show-MyTask` and `shmt` to `Get-MyTask` to provide some sort of backward compatibility.
+ Modified `Get-EmailReminder` to display a warning if the scheduled job is not found.
+ Updated `Enable-EmailReminder` to better handle text output now that default formatting uses ANSI.
+ Updated Pester tests.
+ Help and documentation updates.

## v2.3.0

+ Updates to `Show-MyTask` to better handle long descriptions. (Issue #40 and #41)
+ Modified `Remove-MyTask` to remove by the ID number. (Issue #42)
+ Added online help links.
+ Updated `README.md`.
+ Updated help documentation.
+ Updated Pester tests.

## v2.2.0

+ Fixed bug where overdue tasks were not displaying in red.
+ Added a Table view called DueDate.
+ Minor help updates.
+ Minor updates to `README.md`.

## v2.1.0

+ Renamed `Set-MyTaskPath` to `Set-MyTaskHome` and set the original name as an alias. (Issue #38)
+ Renamed `Get-MyTaskPath` to `Get-MyTaskHome` and set the original name as an alias. (Issue #38)
+ Restructured module to better accommodate Desktop vs Core PSEditions. (Issue #37)
+ Manifest updates.
+ Documentation updates.
+ Updated Pester tests.

## v2.0.0

+ Updated manifest to require PowerShell 5.1 and support for both Desktop and Core PSEditions *Breaking Change*.
+ Added `Get-MyTaskPath` command.(Issue #36)
+ Added a format.ps1xml file for `Get-MyTaskPath`
+ Modified code to determine home folder to use `[Environment]::GetFolderPath([Environment+SpecialFolder]::MyDocuments)` Thank you @kilasuit and @thecliguy. (Issue #35) *Breaking Change*
+ Fixed bug in `Show-Mytask` where year is displayed in 2 digits instead of 4 on Linux platforms.
+ Documentation updates.

## v1.9.0

+ Updated the auto-completer to get task names and enclose in quotes.
+ Fixed bug in `Get-EmailReminder` to determine if the command is supported.
+ Modified `Save-MyTask` to provide more detail when using -Whatif.
+ Added `Get-MyTaskArchive` command.
+ Updated class and commands to better handle OverDue values.
+ Modified `MyTasks.format.ps1xml` to support myTaskArchive type.
+ Updated help.

## v1.8.2

+ Fixed bugs with the email reminder. Typo in a parameter name.
+ Added parameter validation for `-Days` in `Enable-EmailReminder`.
+ Help updates.

## v1.8.0

+ Fixed Task Category bug. Thank you @shaneis. (Issue #26)
+ Revised Pester test for category fix.
+ Updated `README.md`.

## v1.7.0

+ File cleanup for PowerShell Gallery.
+ M.ved aliases to function definitions
+ Added auto completer for `Get-MyTask`, `Set-MyTask`, `Complete-MyTask`, and `Remove-MyTask`. (Issue #32)
+ Updated `Get-MyTask` to support multiple ID numbers. (Issue #31)
+ Help documentation update.
+ General code cleanup.

## v1.6.0

+ Fixed bug saving XML to non-filesystem paths. (Issue #28)
+ Update `Set-MyTaskPath` to support -Passthru. (Issue #25)
+ Updated `Get-EmailReminder` to include additional details. (Issue #18)
+ Fixed CSS bug in email scriptblock.
+ Updated `Enable-EmailReminder` to allow the user to specify the number of days. (Issue #29)
+ Updated `Enable-EmailReminder` to allow the user to specify an alternate path. (Issue #30)
+ Updated Pester tests.
+ Updated documentation.

## v1.5.1

+ Fixed a bug with MyTaskCategory. (Issue #27)

## v1.5.0

+ Added a patch to store the date in an ISO friendly format. (Issue #23 and Issue #22)
+ Added a new function, `Set-MyTaskPath` to allow you to update the task folder. (Issue #20)
+ Updated documentation.

## v1.4.0

+ Added support for PowerShell scheduled job to email upcoming tasks. (Issue #17)
+ Scheduled reminders will be a Windows-only feature.
+ Added timestamp to verbose messages.

## v1.3.1

+ Fixed a bug with `Show-Task` when no entries are found. (Issue #16)
+ Updated README.

## v1.3.0

+ Updated README.
+ Added 'task' as an alias for `New-MyTask`.
+ Added CompletedDate parameter to `Complete-MyTask`. (Issue #13)
+ Added option to complete a task by ID. (Issue #11)
+ Changed default for `Get-MyTask` and `Show-MyTask` to display tasks due in the next 30 days. (Issue #12)
+ Fixed bug in `Show-MyTask` where completed tasks were displaying in red.
+ Made parameters for `New-MyTask` positional. (Issue #14)
+ Updated help documentation.

## v1.2.0

+ Modified files to support PowerShell Core.
+ Update help documentation.
+ Updated `README.md`.
+ Updated manifest.

## v1.1.0

+ Updated verbose messages.
+ Moved class definition to functions script.
+ Revised Pester tests.

## v1.0.4

+ Explicitly added UTF8 encoding when reading and writing XML content.
+ Explicitly added Unicode encoding when reading and writing task category file. (Issue #4)
+ Modified `Show-MyTask` to format as a table with auto-sizing and string streaming.
+ Updated manifest.

## v1.0.3

+ Modified code when creating a new XML fil to specify encoding as UTF-8.
+ Modified `Get-MyTask` to display a warning when trying to display tasks when none have been defined yet.
+ Modified `Show-MyTask` to work under the Windows 10 PowerShell ISE. (Issue #2)
+ Modified `Remove-MyTask` to accept MyTask as an input object. (Issue #3)
+ Updated `README.md`

## v1.0.2

+ Modified `New-MyTask` so that when using the `-Passthru` parameter, it displays the correct ID. (Issue #1)

## v1.0.1

+ Added Pester tests.
+ Fixed a bug with exported variables.

## v1.0.0

+ Added license file.
+ Initial release

## v0.0.16

+ Added help documentation.
+ Revised `Save-MyTask` to save a single task.
+ Revised `Complete-MyTask` to archive a single task.
+ Modified `Get-MyTask` to allow wildcards for -Name.

## v0.0.15

+ Fixed a regular expression bug in `Show-MyTask` that wasn't properly capturing completed tasks.
+ Modified `Show-MyTask` to display completed tasks in green.

Tests/MyTasks.module.tests.ps1
if (Get-Module -Name MyTasks) {
    Remove-Module -Name MyTasks
}

Import-Module -Name "$PSScriptRoot\..\Mytasks.psd1" -Force

Write-Host "These tests are designed for Windows PowerShell" -ForegroundColor yellow
Describe 'MyTasks' {

    $Module = Get-Module -Name MyTasks
    It 'should have 17 functions' {
        $Module.ExportedFunctions.count | Should -Be 16
    }

    It 'should have 10 aliases command' {
        $Module.ExportedAliases.Count | Should -Be 11
    }

    It 'should not export any variables' {
        $Module.ExportedVariables.Count | Should -Be 0
    }

    $modvariables = @(
        @{Variable = 'myTaskArchivePath'}
        @{Variable = 'myTaskCategory'}
        @{Variable = 'mytaskhome'}
        @{Variable = 'mytaskPath'}
    )

    It 'should define a global variable called <Variable>' -TestCases $modVariables {
        param ($Variable)
        {Get-Variable -Name $variable -Scope global} | Should -Not -Throw
    }
    It 'should have 2 formatting xml files' {
        $Module.ExportedFormatFiles.Count | Should -Be 2
    }

    It 'should have an about help topic' {
        {Get-Help about_mytasks} | Should -Not -Throw
    }

    It 'requires PowerShell 5.1' {
        $Module.PowerShellVersion | Should -Be '5.1'
    }
} #describe my module

Describe "Functions" {

    $cmds = @(
        @{Name = 'Disable-EmailReminder'}
        @{Name = 'Enable-EmailReminder'}
        @{Name = 'Get-EmailReminder'}
        @{Name = 'New-MyTask'}
        @{Name = 'Set-MyTask'}
        @{Name = 'Save-MyTask'}
        @{Name = 'Remove-MyTask'}
        @{Name = 'Complete-MyTask'}
        @{Name = 'Show-MyTask'}
        @{Name = 'Get-MyTask'}
        @{Name = 'Add-MyTaskCategory'}
        @{Name = 'Get-MyTaskCategory'}
        @{Name = 'Remove-MyTaskCategory'}
        @{Name = 'Backup-MyTaskFile'}
        @{Name = 'Set-MyTaskHome'}
        @{Name = 'Get-myTaskArchive'}
        @{Name = 'Get-MyTaskHome'}
    )

    It "<Name> has external help defined with at least one example" -TestCases $cmds {
        param($Name)
        $help = Get-Help $Name
        $help.Description | Should -Not -BeNullOrEmpty
        $help.Examples | Should -Not -BeNullOrEmpty
    }

    It "<Name> has an online help link" -TestCases $cmds {
        param($Name)
        $help = Get-Help $Name
        $help.relatedLinks.navigationlink.uri -match "http:" | Should -Not -BeNullOrEmpty
    }

    Context 'Enable-EmailReminder' {
        $cmd = Get-Command -Name Enable-EmailReminder
        $params = "TO", "TASKCREDENTIAL"
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
    }

    Context 'New-MyTask' {
        $cmd = Get-Command -Name New-MyTask
        $params = "NAME", "CATEGORY"
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
        $params = 'NAME'.'DUEDATE'.'DESCRIPTION'.'CATEGORY'
        Foreach ($item in $params) {
            It "should have a $item parameter that accepts pipeline input by name" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).ValueFromPipelineByPropertyName | Should -be $True
            }
        }
    }
    Context 'Set-MyTask' {
        $cmd = Get-Command -Name Set-MyTask
        $params = "NAME"
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }

        $params = "NAME"
        Foreach ($item in $params) {
            It "should have a $item parameter that accepts pipeline input by name" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).ValueFromPipelineByPropertyName | Should -be $True
            }
        }
    }
    Context 'Save-MyTask' {
        $cmd = Get-Command -Name Save-MyTask
        $params = "TASK"
        Foreach ($item in $params) {
            It "should have a $item parameter that accepts pipeline input by value" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).ValueFromPipeline| Should -be $True
            }
        }
    }
    Context 'Remove-MyTask' {
        $cmd = Get-Command -Name Remove-MyTask
        $params = "NAME", 'INPUTOBJECT'
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
    }
    Context 'Complete-MyTask' {
        $cmd = Get-Command -Name Complete-MyTask
        $params = "NAME", 'ID'
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
    }

    Context 'Add-MyTaskCategory' {
        $cmd = Get-Command -Name Add-MyTaskCategory
        $params = "CATEGORY"
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
        Foreach ($item in $params) {
            It "should have a $item parameter that accepts pipeline input by value" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).ValueFromPipeline| Should -be $True
            }
        }
    }

    Context 'Remove-MyTaskCategory' {
        $cmd = Get-Command -Name Remove-MyTaskCategory
        $params = "CATEGORY"
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
        Foreach ($item in $params) {
            It "should have a $item parameter that accepts pipeline input by value" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).ValueFromPipeline| Should -be $True
            }
        }
    }

    Context 'Set-MyTaskHome' {
        $cmd = Get-Command -Name Set-MyTaskHome
        $params = "PATH"
        Foreach ($item in $params) {
            It "should have a mandatory $item parameter" {
                ($cmd.parameters["$item"].Attributes).where( {$_.typeid.name -match 'parameterAttribute'}).Mandatory | Should -be $True
            }
        }
    }

    Context 'Get-MyTaskArchive' {

    }

    Context 'Get-MyTaskHome' {

    }

} #describe functions

functions/emailfunctions.ps1
Function Enable-EmailReminder {
    #this function requires the PSScheduledJob module
    [cmdletbinding(SupportsShouldProcess)]
    [OutputType("None")]

    Param(
        [Parameter(Position = 0, HelpMessage = "What time do you want to send your daily email reminder?")]
        [ValidateNotNullOrEmpty()]
        [datetime]$Time = "8:00AM",
        [Parameter(HelpMessage = "What is your email server name or address?")]
        [ValidateNotNullOrEmpty()]
        [string]$SMTPServer = $PSEmailServer,
        [Parameter(Mandatory, HelpMessage = "Enter your email address")]
        [ValidateNotNullOrEmpty()]
        [ValidatePattern("\S+@*.\.\w{2,4}")]
        [string]$To,
        [Parameter(HelpMessage = "Enter the FROM email address. If you don't specify one, the TO address will be used.")]
        [ValidatePattern("\S+@*.\.\w{2,4}")]
        [string]$From,
        [Parameter(HelpMessage = "Include if you need to use SSL?")]
        [switch]$UseSSL,
        [Parameter(HelpMessage = "Specify the port to use for your email server")]
        [ValidateNotNullOrEmpty()]
        [int32]$Port = 25,
        [Parameter(HelpMessage = "Specify any credential you need to authenticate to your mail server.")]
        [PSCredential]$MailCredential,
        [Parameter(HelpMessage = "Send an HTML body email")]
        [switch]$AsHtml,
        [ValidateNotNullOrEmpty()]
        [ValidateScript( { $_ -gt 0 })]
        [int]$Days = 3,
        [Parameter(Mandatory, HelpMessage = "Re-enter your local user credentials for the scheduled job task")]
        [ValidateNotNullOrEmpty()]
        [PSCredential]$TaskCredential,
        [ValidateNotNullOrEmpty()]
        [string]$TaskPath = $mytaskHome
    )
    Begin {
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN  ] Starting $($myinvocation.mycommand)"
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN  ] Using these mail parameters:"

        if (-Not $From) {
            $From = $To
        }
        $hash = @{
            To         = $To
            From       = $From
            SMTPServer = $SMTPServer
            Port       = $Port
        }
        if ($MailCredential) {
            $hash.Add("Credential", $MailCredential)
        }
        If ($UseSSL) {
            $hash.add("UseSSL", $True)
        }

        $hash | Out-String | Write-Verbose
        #define the job scriptblock
        $sb = {
            Param([hashtable]$Hash, [int]$Days, [string]$myPath)
            #uncomment Write-Host lines for troubleshooting
            #$PSBoundParameters | out-string | write-host -ForegroundColor cyan
            #get tasks for the next 3 days as the body

            Set-MyTaskPath -Path $myPath
            #get-variable mytask* | out-string | Write-Host
            Write-Host "[$((Get-Date).ToString())] Getting tasks for the next $days days."
            $data = Get-MyTask -Days $Days
            if ($data) {
                if ($hash.BodyAsHTML) {
                    Write-Host "[$((Get-Date).ToString())] Sending as HTML" -ForegroundColor green
                    #css to be embedded in the html document
                    $head = @"
<Title>Upcoming Tasks</Title>
<style>
body { background-color:rgb(199, 194, 194);
    font-family:Tahoma;
    font-size:12pt; }
td, th { border:1px solid black;
            border-collapse:collapse; }
th { color:white;
        background-color:black; }
table, tr, td, th { padding: 2px; margin: 0px }
tr:nth-child(odd) {background-color: lightgray}
table { width:95%;margin-left:5px; margin-bottom:20px;}
.alert {color: red ;}
.warn {color:#ff8c00;}
</style>
<br>
<H1>My Tasks</H1>
"@
                    [xml]$html = $data | ConvertTo-Html -Fragment

                    #parse html to add color attributes
                    for ($i = 1; $i -le $html.table.tr.count - 1; $i++) {
                        $class = $html.CreateAttribute("class")
                        #check the value of the percent free memory column and assign a class to the row
                        if ($html.table.tr[$i].td[4] -eq 'True') {
                            $class.value = "alert"
                            $html.table.tr[$i].Attributes.Append($class) | Out-Null
                        }
                        elseif ((($html.table.tr[$i].td[3] -as [DateTime]) - (Get-Date)).totalHours -le 24 ) {
                            $class.value = "warn"
                            $html.table.tr[$i].Attributes.Append($class) | Out-Null
                        }
                    }

                    $Body = ConvertTo-Html -Body $html.InnerXml -Head $head | Out-String

                }
                else {
                    Write-Host "[$((Get-Date).ToString())] Sending as TEXT" -ForegroundColor Green
                    # 10/14/2020 Modified to explictly select properties because
                    # default formatting uses ANSI which distorts the converted output.
                    $body = $data | Select-Object -property ID, Name, Description, DueDate, OverDue | Format-Table | Out-String
                }
            }
            else {
                Write-Warning "No tasks found due in the next $days days."
                #bail out
                return
            }
            $hash.Add("Body", $body)
            $hash.Add("Subject", "Tasks Due in the Next $days Days")
            $hash.Add("ErrorAction", "Stop")
            Try {
                Send-MailMessage @hash
                #if you receive the job I wanted to display some sort of result
                Write-Output "[$((Get-Date).ToString())] Message ($($hash.subject)) sent to $($hash.to) from $($hash.from)"
            }
            Catch {
                throw $_
            }
        } #define scriptblock
    } #begin

    Process {
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Testing for an existing job"
        Try {
            $job = Get-ScheduledJob -Name myTasksEmail -ErrorAction stop
            if ($job) {
                Write-Warning "An existing mail job was found. Please remove it first with Disable-EmailReminder and try again."
                #bail out
                return
            }
        }
        Catch {
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] No existing job found"
        }
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Validating Requirements"
        if ((Get-Module PSScheduledJob) -And (($PSVersionTable.Platform -eq 'Win32NT') -OR ($PSVersionTable.PSEdition -eq 'Desktop'))) {

            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Creating a Daily job trigger for $($Time.TimeofDay)"
            $trigger = New-JobTrigger -Daily -At $Time

            $opt = New-ScheduledJobOption -RunElevated -RequireNetwork
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Registering the scheduled job"
            if ($AsHtml) {
                $hash.Add("BodyAsHTML", $True)
            }
            #hash table of parameters to splat Register-ScheduledJob
            $regParams = @{
                ScriptBlock        = $sb
                Name               = "myTasksEmail"
                Trigger            = $Trigger
                ArgumentList       = $hash, $Days, $TaskPath
                MaxResultCount     = 5
                ScheduledJobOption = $opt
                Credential         = $TaskCredential
            }
            $regParams | Out-String | Write-Verbose
            Register-ScheduledJob  @regParams
        }
        else {
            Write-Warning "This command requires the PSScheduledJob module on a Windows platform."
        }
    } #process

    End {
        Write-Verbose "[$((Get-Date).TimeofDay) END    ] Ending $($myinvocation.mycommand)"
    } #end

} #close Enable-MailReminder

Function Disable-EmailReminder {
    [cmdletbinding(SupportsShouldProcess)]
    [OutputType("None")]

    Param()
    Begin {
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN  ] Starting $($myinvocation.mycommand)"
    } #begin

    Process {
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Processing"
        if ((Get-Module PSScheduledJob) -And (($PSVersionTable.Platform -eq 'Win32NT') -OR ($PSVersionTable.PSEdition -eq 'Desktop'))) {
            Try {
                if (Get-ScheduledJob -Name myTasksEmail -ErrorAction stop) {
                    #The cmdlet appears to ignore -WhatIf so I'll handle it myself
                    if ($PSCmdlet.ShouldProcess("myTasksEmail")) {
                        Unregister-ScheduledJob -Name myTasksEmail -ErrorAction stop
                    } #should process
                } #if task found
            }
            Catch {
                Write-Warning "Can't find any matching scheduled jobs with the name 'myTasksEmail'."
            }
        }
        else {
            Write-Warning "This command requires the PSScheduledJob module on a Windows platform."
        }
    } #process

    End {
        Write-Verbose "[$((Get-Date).TimeofDay) END    ] Ending $($myinvocation.mycommand)"
    } #end

} #close Disable-EmailReminder

Function Get-EmailReminder {
    [cmdletbinding()]
    Param ()

    Begin {
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN  ] Starting $($myinvocation.mycommand)"

    } #begin

    Process {
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Getting scheduled job myTasksEmail"
        if ((Get-Module PSScheduledJob -ListAvailable) -And (($PSVersionTable.Platform -eq 'Win32NT') -OR ($PSVersionTable.PSEdition -eq 'Desktop'))) {
            Try {
                $t = Get-ScheduledJob myTasksEmail -erroraction Stop
            }
            Catch {
                Write-Warning "Could not find the Scheduled Job myTasksEmail"
            }

            if ($t) {
                $hash = $t.InvocationInfo.Parameters[0].where( { $_.name -eq "argumentlist" }).value

                Try {
                    #get the last run
                    Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Getting last job run"
                    $last = Get-Job -Name $t.name -Newest 1 -ErrorAction stop
                }
                Catch {
                    $last = [PSCustomObject]@{
                        PSEndTime = "11/30/1999" -as [datetime]
                        State     = "The task has not yet run"
                    }
                }
                [pscustomobject]@{
                    Task       = $t.Name
                    Frequency  = $t.JobTriggers.Frequency
                    At         = $t.JobTriggers.at.TimeOfDay
                    To         = $hash.To
                    From       = $hash.From
                    MailServer = $hash.SMTPServer
                    Port       = $hash.Port
                    UseSSL     = $hash.UseSSL
                    AsHTML     = $hash.BodyAsHTML
                    LastRun    = $last.PSEndTime
                    LastState  = $last.State
                    Started    = $last.psBeginTime
                    Ended      = $last.psEndTime
                    Result     = $last.output
                    Enabled    = $t.Enabled
                    Errors     = $last.Errors
                    Warnings   = $last.warnings
                }
            }
        }
        else {
            Write-Warning "This command requires the PSScheduledJob module on a Windows platform."
        }
    } #process

    End {
        Write-Verbose "[$((Get-Date).TimeofDay) END    ] Ending $($myinvocation.mycommand)"

    } #end

} #close Get-EmailReminder